cluster_management/src/main/java/com/pinterest/rocksplicator/eventstore/ExternalViewLeaderEventsLoggerImpl.java

```diff
@@ -0,0 +1,322 @@
+        Map<String, String> hostToState = externalView.getStateMap(partitionName);
+        // We don't know who was the leader that went down...
+        // all we know is that a leader is not available for
+        // this partition, unless we find a host with leader state.
+        boolean leaderFound = false;
+        for (Map.Entry<String, String> entry : hostToState.entrySet()) {
+          String host = entry.getKey();
+          String state = entry.getValue();
+
+          if (ExternalViewUtils.StateUtils.isStateAnyKindOfLeader(state)) {
+            //Since this host is disabled, whatever is the state, we consider this to be down
+            leaderFound = true;
+            if (disabledHosts.contains(host)) {
+              if (previouslyKnownLeader == null
+                  || previouslyKnownLeader.state == LeaderState.LEADER_UP
+                  || (previouslyKnownLeader.state == LeaderState.LEADER_DOWN
+                          && (previouslyKnownLeader.leaderInstanceId == null || !host
+                  .equals(previouslyKnownLeader.leaderInstanceId)))) {
+                /**
+                 * We publish the event as LEADER_DOWN if
+                 * either previously there was no information about this partition
+                 * or previously the known leader was up
+                 * or previously the leader host was not known
+                 * or if previous leader was known, it has changed now.
+                 */
+                LOGGER.error(String
+                    .format("LEADER_DOWN (Published) resource:%s, partition=%s, leader:%s",
+                        resourceName, partitionName, host));
+                resourceLeaderCache.put(partitionName, new Leader(host, LeaderState.LEADER_DOWN));
+                publishStatus(resourceName, partitionName, host, LeaderState.LEADER_DOWN);
+              } else {
+                // Avoid posting duplicate event.
+                LOGGER.error(String
+                    .format("LEADER_DOWN (Duplicate/Skipped) resource:%s, partition=%s, leader:%s",
+                        resourceName, partitionName, host));
+                doNothing();
+              }
+            } else {
+              if (previouslyKnownLeader == null
+                  || previouslyKnownLeader.state == LeaderState.LEADER_DOWN
+                  || (previouslyKnownLeader.state == LeaderState.LEADER_UP
+                          && (previouslyKnownLeader.leaderInstanceId == null || !host
+                  .equals(previouslyKnownLeader.leaderInstanceId)))) {
+                /**
+                 * We publish the event as LEADER_DOWN if
+                 * either previously there was no information about this partition
+                 * or previously the known leader was up
+                 * or previously the leader host was not known
+                 * or if previous leader was known, it has changed now.
+                 */
```

Comment:
Can you update comments here to explain the LEADER_UP state?

Reply:
If the spectator has just come up, it won't have cached previous leader known...
if previous event we found that some leader was up state..
or if (it was known to be down) ... but we didn't know which host was leader... hence update..
or if the known leader that is down has changed... In all of the above cases... there is a change in either known state or actual leader that is down... hence we publish the event...

Basic idea is to only report each of the state/leader only once, the first time it is encountered... and subsequently supress same kind of event... as it is immaterial. Next time we publish is when there is a material change in either the status of known leader or the leader itself changed...